### PR TITLE
Change EcdsaSig hash type deser in psbt

### PR DESCRIPTION
**Pull request by sanket1729**
Changes the parsing behavior in PSBT on non-standard sighash types to give an explicit error, rather than silently mangling the parsed value

### CODE REVIEW

File: src/util/psbt/serialize.rs [hunk 1]
```diff
@@ -201,20 +206,15 @@ impl Serialize for schnorr::SchnorrSig  {
...
-                    .map_err(|_| encode::Error::ParseFailed("Invalid Schnorr signature"))?;
-                    Ok(schnorr::SchnorrSig{ sig, hash_ty: SchnorrSigHashType::Default })
+        match schnorr::SchnorrSig::from_slice(&bytes) {
+            Ok(sig) => Ok(sig),
+            Err(schnorr::SchnorrSigError::InvalidSighashType(flag)) => {
+                Err(encode::Error::from(psbt::Error::NonStandardSigHashType(flag as u32)))
             }
-            _ => Err(encode::Error::ParseFailed("Invalid Schnorr signature len"))
+            Err(schnorr::SchnorrSigError::InvalidSchnorrSigSize(_)) =>
+                Err(encode::Error::ParseFailed("Invalid Schnorr signature len")),
```
Comment (Kixunil) [inline src/util/psbt/serialize.rs:215]: Nit: I'd prefer not to abbreviate "length" in an error message.

File: src/util/psbt/serialize.rs [hunk 2]
```diff
@@ -90,25 +90,35 @@ impl Deserialize for secp256k1::PublicKey {
...
+        // would use check the signature assuming sighash_u32 as `0x01`.
+        match EcdsaSig::from_slice(&bytes) {
+            Ok(sig) => Ok(sig),
+            Err(EcdsaSigError::EmptySignature) =>
+                Err(encode::Error::ParseFailed("Empty partial signature data")),
+            Err(EcdsaSigError::NonStandardSigHashType(flag)) =>
+                Err(encode::Error::from(psbt::Error::NonStandardSigHashType(flag))),
+            Err(EcdsaSigError::Secp256k1(..)) =>
+                Err(encode::Error::ParseFailed("Invalid Ecdsa signature")),
+            Err(EcdsaSigError::HexEncoding(..)) => unreachable!("Decoding from slice, not hex")
```
Comment (Kixunil) [inline src/util/psbt/serialize.rs:120]: Unrelated but this is why I don't like catch-all error types.

**Review by Kixunil — COMMENTED** [review]
Looks correct. How hard would it be to introduce a new type? I guess not even urgent for Taproot, right?

File: src/blockdata/transaction.rs [hunk 3]
```diff
@@ -758,6 +758,10 @@ impl EcdsaSigHashType {
      ///
      /// **Note**: this replicates consensus behaviour, for current standardness rules correctness
      /// you probably want [Self::from_u32_standard].
+     /// This might cause unexpected behavior because it does not roundtrip. That is,
+     /// `EcdsaSigHashType::from_u32_consensus(n) as u32 != n` for non-standard values of
+     /// `n`. While verifying signatures, the user should retain the `n` and use it compute the
+     /// signature hash message.
      pub fn from_u32_consensus(n: u32) -> EcdsaSigHashType {
```
[the remark this answers is not in the document]
Comment (sanket1729) [reply src/blockdata/transaction.rs:765]: I think *if* people are dealing with non-standard stuff, they better know what they are doing. So,  I don't think we should concern ourselves by adding more code to safeguard non-standard users. 

It is not an effort to introduce a new type, but I think I would be more comfortable with less code bloat about a code that no one is using. Document the non-standard stuff, and if someone wants it they can PR one and raise an issue.

File: src/util/ecdsa.rs [hunk 4]
```diff
@@ -451,9 +451,10 @@ impl EcdsaSig {
     /// Serialize EcdsaSig
     pub fn to_vec(&self) -> Vec<u8> {
         // TODO: add support to serialize to a writer to SerializedSig
-        let mut ser_sig = self.sig.serialize_der().to_vec();
-        ser_sig.push(self.hash_ty.as_u32() as u8);
-        ser_sig
+        let mut buf = Vec::with_capacity(72);
+        buf.extend(self.sig.serialize_der().iter());
+        buf.push(self.hash_ty as u8);
+        buf
```
[the remark this answers is not in the document]
Comment (sanket1729) [reply src/util/ecdsa.rs]: MSRV issues. 

```
error[E0599]: no method named `copied` found for type `std::slice::Iter<'_, u8>` in the current scope
   --> src/util/ecdsa.rs:454:41
    |
454 |         self.sig.serialize_der().iter().copied()
    |                                         ^^^^^^
```

[the remark this answers is not in the document]
Comment (sanket1729) [reply src/util/ecdsa.rs]: did a manual map and copy instead

**Review by dr-orlovsky — DISMISSED** [review]
ACK efbeba0a56088078e14f89fa26a03884a5460f8a

**Comment by dr-orlovsky** [comment]
Needs rebase after #779 merge

File: src/blockdata/transaction.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply src/blockdata/transaction.rs:765]: We still need non-standard to validate consensus/process valid block, right? Couldn't it be a problem?

[the remark this answers is not in the document]
Comment (sanket1729) [reply src/blockdata/transaction.rs:765]: Yes, we need it. Our code for `legacy_signature_hash` does take input a sighash_u32 that we expect the user to provide correctly! 

However, our code for bip143 segwitv0 sighash does not support signature verification of non-standard transactions! This is a bug and should be addressed independently.

**Review by dr-orlovsky — APPROVED** [review]
ACK abe52f681bc5d84b74fa1299f1b9751b0a222ea6

**Comment by dr-orlovsky** [comment]
@Kixunil let's get this closed and ship taproot! 🥇

**Review by apoelstra — APPROVED** [review]
ACK abe52f681bc5d84b74fa1299f1b9751b0a222ea6

@sanket1729 can you change the description from "borderline trivial" to "Changes the parsing behavior in PSBT on non-standard sighash types to give an explicit error, rather than silently mangling the parsed value"?

My first thought was to NACK this PR because we didn't have agreement on what the correct behavior was, but after studying it, I realized that our current behavior is definitely wrong and that this is at least an improvement.

**Comment by sanket1729** [comment]
@apoelstra changed the description.

**Comment by dr-orlovsky** [comment]
We already have 2 ACKs, but I assume it would be nice to hear from @Kixunil reviewing this previously that all his comments are addresses before merging this

File: src/util/psbt/serialize.rs [hunk 5]
```diff
@@ -194,20 +205,15 @@ impl Serialize for schnorr::SchnorrSig  {
...
+        match schnorr::SchnorrSig::from_slice(&bytes) {
+            Ok(sig) => Ok(sig),
+            Err(schnorr::SchnorrSigError::InvalidSighashType(flag)) => {
+                Err(encode::Error::from(psbt::Error::NonStandardSigHashType(flag as u32)))
             }
-            _ => Err(encode::Error::ParseFailed("Invalid Schnorr signature len"))
+            Err(schnorr::SchnorrSigError::InvalidSchnorrSigSize(_)) =>
+                Err(encode::Error::ParseFailed("Invalid Schnorr signature length")),
+            Err(schnorr::SchnorrSigError::Secp256k1(..)) =>
+                Err(encode::Error::ParseFailed("Invalid Schnorr signature")),
```
Comment (Kixunil) [inline src/util/psbt/serialize.rs:216]: This could've been `map_err` which would communicate the intent a bit better. Only realized it now.

**Review by Kixunil — APPROVED** [review]
ACK abe52f681bc5d84b74fa1299f1b9751b0a222ea6

Go ahead and merge this, we can make a trivial followup issue/PR to rewrite to `map_err`

File: src/util/psbt/serialize.rs (hunk 5 above)
Comment (dr-orlovsky) [reply src/util/psbt/serialize.rs:216]: Issue: https://github.com/rust-bitcoin/rust-bitcoin/issues/793

**Pull request merged**